substratevm/src/com.oracle.svm.configure/src/com/oracle/svm/configure/trace/AccessAdvisor.java

```diff
@@ -78,6 +78,11 @@
 
         internalCallerFilter.addOrGetChildren("org.graalvm.compiler.**", RuleNode.Inclusion.Exclude);
         internalCallerFilter.addOrGetChildren("org.graalvm.libgraal.**", RuleNode.Inclusion.Exclude);
+
+        // For serializations
+        internalCallerFilter.addOrGetChildren("java.io.ObjectInputStream", RuleNode.Inclusion.Include);
+        internalCallerFilter.addOrGetChildren("java.io.ObjectOutputStream", RuleNode.Inclusion.Include);
+        internalCallerFilter.addOrGetChildren("java.io.ObjectStreamClass", RuleNode.Inclusion.Include);
```

Comment:
Why are these necessary? These rules should cause reflective accesses from the various stream classes to end up in the reflection configuration, but these should be considered implementation details of serialization/deserialization on native-image and we would already register accessible fields separately.

Reply:
These rules are included because not all serialization/deserialization calls go to MethodAccessorGeneretor.generateSerializationConstructor. While serializing/deserializing Arrays, Enums, Externalizable classes and classes without proper super constructors, MethodAccessorGeneretor.generateSerializationConstructor will not be invoked (more details can be seen at `private ObjectStreamClass(final Class<?> cl)` ) .
If these rules should not get included, I need to install a new method break point on  `private ObjectStreamClass(final Class<?> cl)` to cover all target class of serialization/deserialization.

Reply:
Thanks!
Shouldn't the serialization constructor and other members of an `ObjectStreamClass` for a specific `Class<?>` always be the same? If so, then we could (through reflection itself, or via JVMCI) instantiate a specific `ObjectStreamClass` during the image build, and read its `cons`, `fields`, `readObjectMethod`, ... fields and register these members for reflective access. But actually, maybe registering the members is not even necessary, just having the `ObjectStreamClass` instances on the image heap could be enough so that everything will be readily accessible at runtime. Since all accesses go through static method `ObjectStreamClass.lookup`, we could substitute that method to access the collection of `ObjectStreamClass` instances created during the image build. The same probably applies to `ObjectStreamClass.getReflector`. We can then probably substitution-delete the `Caches` class because there is no longer a need for a mutable _cache_.

Reply:
I experimented a bit with the code today to try (most of) the changes I described: https://github.com/oracle/graal/pull/2794/commits/f531e52ca6170ae5b735e54d3b77a53a633a8637
Preparing the collection of `ObjectStreamClass` objects during the image build works, but then the generated constructors generate the wrong type of object at image runtime, such as just `java.lang.Object` instead of the specific subclass. Invoking the constructor at image build time in the debugger instantiates the correct class, however. Maybe this is what you hinted at earlier, but I don't quite understand why `parameterTypes`, `checkedExceptions`, `modifiers` and `targetConstructorClass` would be different between the run with the agent and with the image build.

Reply:
I've run the tests in [tests.zip](https://github.com/oracle/graal/files/5129056/tests.zip) with your patch and here's the result:

1. testCustomizedClassSerialize.sh crashed at run time. This test serializes and deserializes a class with multiple types of fields. The last stack traces are:
```
Stacktrace Stage 2:

  SP 00007ffd299e91f0 IP 0000000000509fe9  com.oracle.svm.core.code.CodeInfo@0x7f4d58f04860 name = image code
  SP 00007ffd299e9250 IP 00000000005045a6  [image code] java.io.ObjectStreamClass.getObjFieldValues(ObjectStreamClass.java:1413)
  SP 00007ffd299e9250 IP 00000000005045a6  [image code] java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1544)
```
  This crash suggests storing `ObjectStreamClass` alone at build time in the native heap is not sufficient. `ObjectStreamClass` 's inner classes are also required to get stored at build time. 

2. testSerializeArrayList.sh failed with an exception. This test serializes and deserializes the `ArrayList` class. The exception message is:
```
Caused by: java.lang.InstantiationException: Cannot instantiate class java.util.AbstractList
        at java.lang.reflect.Constructor.newInstance(Constructor.java:490)
        at java.io.ObjectStreamClass.newInstance(ObjectStreamClass.java:1092)
        at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2150)
```
The `ArrayList` class doesn't have so complicated fields as the previous one, so it goes further at runtime. This failure suggests the first unserializable super class of the serialization target class is not prepared at build time. This can be fixed by registering the `ObjectStreamClass`'s `cons` field into `reflectionData` at build time.

3. testDeserializeStream.sh passed because it deserializes a class of float array, which doesn't have field nor serialization constructor.

I think storing `ObjectStreamClass` in native heap is doable for general cases, but it will fail when `parameterTypes`, `checkedExceptions`, `modifiers` and `targetConstructorClass` are different between agent run time and image build time. This happens when two same name classes with different `parameterTypes`, `checkedExceptions`, `modifiers` and `targetConstructorClass` are loaded by different classloaders at agent run time, however, native-image can't load such classes with run time's classloaders at build time. The test case `testDeserializeMultiClassloader.sh` in test.zip shows such a case. If any two configuration items have the same `targetClass` but different `parameterTypes`, `checkedExceptions`, `modifiers` or `targetConstructorClass`, we know such a multi-classloaders case happens and an error can be reported.

Reply:
My patch with build-time `ObjectStreamClass` is incomplete so failing tests are expected. I have looked into the object instantiation problems further: during the image build, we substitute reflective `newInstance` with a generated `ReflectiveNewInstanceMethod` that creates and initializes the object directly without the serialization constructor accessor. Because the serialization constructor is injected into a superclass, our substitution instantiates the wrong class and uses the wrong constructor. I managed to get around that in my updated changes in 72411338a8c6c719ca84ad84a2b88b7338282982, although I still see an aliasing bug between serialization constructors that can break deserialization.
The next issue I encountered is that creating an `ObjectStreamClass` ahead of time will record the object field offsets in `FieldReflector` during the image build, which can be different from the offsets at image runtime. This can be fixed too of course, but this and other cases probably make creating an `ObjectStreamClass` at runtime the simpler option. With that variant, the `ObjectStreamClass.Caches` members should be cleared via `@RecomputeFieldValue(kind = NewInstance)`. This also goes for members of `ObjectInputStream.Caches` and `ObjectOutputStream.Caches` and possibly other classes.

Concerning the multiple class loaders, is it really a common issue to have multiple class loaders that load different classes with the same name for serialization where `parameterTypes`, `checkedExceptions`, `modifiers` or `targetConstructorClass` differ? Wouldn't it be preferable to include a hash (or just size) or the `serialVersionUID` of a (de)serializable class in an agent-generated configuration and either print a warning or abort if they mismatch during the image build? (I believe this should be optional in the configuration file)

Reply:
>Concerning the multiple class loaders, is it really a common issue to have multiple class loaders that load different classes with the same name for serialization where parameterTypes, checkedExceptions, modifiers or targetConstructorClass differ? 
It is not common, but we must assure the correctness. One approach is to report multi-classloader usage issue at Agent run time, then we don't need the extra information in configuration file to check at build time, only serialization target class should work.

Reply:
Output from the agent is likely to remain unseen among the output of the observed application, so I would avoid that.

I still don't quite understand how keeping these four values around helps to identify when different classes are loaded by multiple class loaders. Unless their class hierarchy differs, wouldn't similar classes have identical serialization constructors? For example, if two class loaders load different versions of a class, one of which has an extra field, would that still be detected? And if so, what about when two class loaders load an identical class? I think that is quite possible in very modular applications. Since it is possible that the class path for the image build is different from the application running with the agent, shouldn't a check like this apply to the serialized class itself in the form of a hash code or its `serializationVersionUID`?

Reply:
SerializationConstructorAccessor class differs only when class hierarchy changes, while it is not affected by the change of class contents.  

JDK caches `ObjectStreamClass` instances in `ObjectStreamClass.Caches.localDescs` which takes the serialization target class' `identityHashcode` as part of its hash value. Therefore, any different class instances (e.g. same name different hierarchy classes, identical classes loaded by different classloaders) will have their own `ObjectStreamClass` instances, leading to a creation of new `SerializationConstructorAccessor` class. 

We wouldn't support the same name different hierarchy case. Using a checksum like hashcode or `serializationVersionUID` in configuration can help to tell the different and report issue at build time.

We can allow the multiple loading of identical class case. Sharing the same `ObjectStreamClass` should be fine for them. But hashcode is not suitable for this case, because classes loaded by different classloaders have different hashcodes.  But their `serializationVersionUID` should be same.

So `serializationVersionUID` is good to use as checksum, but not directly use the class's `serializationVersionUID` field, because user can set an unchanged `serializationVersionUID` value event the contents or hierarchy are changed. We can apply the `serializationVersionUID`'s calculation algorithm to set a checksum and verify it through image build time and image run time.

I think creating `ObjectStreamClass` at image run time is safer and simpler. We can use serialization target class to create and cache `SerializationConstructorAccessor` class at build time, and use `serializationVersionUID` algorithmed checksum to verify the class is identical from Agent run time to image build time and image run time. 

I'll make this change if it sounds good to you.

Reply:
> I think creating `ObjectStreamClass` at image run time is safer and simpler. We can use serialization target class to create and cache `SerializationConstructorAccessor` class at build time, and use `serializationVersionUID` algorithmed checksum to verify the class is identical from Agent run time to image build time and image run time.

Yes, that sounds good to me. I agree about using the hash generated by `computeDefaultSUID` because I also think that users often don't maintain the `serialVersionUID` field properly when modifying the class. I would avoid the term "version UID" in though because it sounds confusing. Instead, I would call it something like `classHashCode`. The configuration entry would then look like this the following. If the hash code doesn't match the one computed during the image build, then the build should fail. It should be possible to not provide a hash code in the configuration and pass the build however.
```
[
  {  "name":"com.oracle.ClassA", "classHashCode":-6949620193861773680  },
  {  "name":"com.oracle.ClassB", "classHashCode":1215806704639619402  }
]
```

Reply:
Maybe a simply `checksum` would be good.

Reply:
Yes, `checksum` is even better. I suppose _hash code_ sounds like it is related to `Object.hashCode`.

Reply:
I have committed a modified version. In this version, the `checksum` is calculated from the serialization constructor class for class hierarchy verification. As the `ObjectStreamClass` instance is created at image run time, the only thing required to cache at build time is `SerializationConstructorAccessor` class which is affected by serialization target class and serialization constructor class. Therefore, given the same serialization target class name, checksum on constructor class name is sufficient to distinguish `SerializationConstructorAccessor` class.

Two options `-H:+AllowUncheckedSerialization` and `-R:+AllowUncheckedSerialization` are provided to skip checking checksum if user provides a serialization configuration without setting checksum.